plugins/ccgx/fu-ccgx-hpi.c

```diff
@@ -0,0 +1,1633 @@
+		}
+	}
+
+	fu_ccgx_hpi_clear_all_event (device, hpi_handle, HPI_CMD_SETUP_EVENT_CLEAR_TIME_MS, &error_local);
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_hpi_get_fw_version (FuDevice *device,
+			    CyHPIHandle *hpi_handle,
+			    guint32 start_row,
+			    guint32 row_size,
+			    PDFWAppVersion *fw_version,
+			    GError **error)
+{
+	guint32 version_row_index = 0;
+	guint32 version_row_offset = 0;
+	gint32 version_row_num = 0;
+	g_autofree guint8 *row_buffer = NULL;
+
+	version_row_index = CCGX_APP_VERSION_OFFSET / row_size;
+	version_row_offset = CCGX_APP_VERSION_OFFSET % row_size;
+
+	version_row_num = start_row + version_row_index;
+	row_buffer = g_malloc0 (row_size);
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+	g_return_val_if_fail (fw_version != NULL, FALSE);
+	g_return_val_if_fail (row_buffer != NULL, FALSE);
+
+	if (!fu_ccgx_hpi_cmd_read_flash (device, hpi_handle, version_row_num, row_buffer, row_size, error)) {
+		g_prefix_error (error, "get fw version error: ");
+		return FALSE;
+	}
+
+	fw_version->val= (guint32) (row_buffer[version_row_offset + 3] << 24 |
+					row_buffer[version_row_offset + 2] << 16 |
+					row_buffer[version_row_offset + 1] << 8 |
+					row_buffer[version_row_offset + 0]);
+	return TRUE;
+ }
+
+/**
+ * fu_ccgx_hpi_cmd_get_device_data:
+ * @device: #FuDevice
+ * @hpi_handle HPI handle
+ * @device_data device data information
+ * @error: a #GError or %NULL
+ *
+ * Get device data (version, fw mode, num of ports, silicon ID) from PD I2C Device
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_hpi_cmd_get_device_data (FuDevice *device,
+				 CyHPIHandle *hpi_handle,
+				 PDDeviceData *device_data,
+				 GError **error)
+{
+	guint32 device_versions[6] = {0};
+	guint8 device_mode = 0;
+	guint16 silicon_id = 0;
+	CCGxPartInfo *ccgx_info = NULL;
+	g_autofree guint8 *row_buffer = NULL;
+	guint32 row_size = 0;
+	guint32 row_max = 0;
+	guint32 fw_meta_row_num;
+	guint32 fw_meta_offset = 0;
+	guint8 fw_index = 0;
+	CCGxMetaData *metadata = NULL;
+	PDFWAppVersion *fw_version = NULL;
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+	g_return_val_if_fail (device_data != NULL, FALSE);
+
+	if (!fu_ccgx_hpi_get_device_mode (device,hpi_handle,&device_mode,error)) {
+		return FALSE;
+	}
+
+	device_data->fw_mode = (FWMode)(device_mode & 0x03);
+
+	if (!fu_ccgx_hpi_get_silicon_id (device,hpi_handle,&silicon_id,error)) {
+		return FALSE;
+	}
+
+	device_data->silicon_id = (guint16)silicon_id;
+
+	device_data->current_version.val = 0;
+	if (device_data->fw_mode != FW_MODE_BOOT) {
+		if (!fu_ccgx_hpi_get_device_version (device, hpi_handle, device_versions, error)) {
+			return FALSE;
+		}
+		device_data->fw_version[FW_MODE_FW1].val = device_versions[3];
+		device_data->fw_version[FW_MODE_FW2].val = device_versions[5];
+		device_data->current_version.val = device_data->fw_version[device_data->fw_mode].val ;
+	} else {
+		g_warning("device in boot mode");
```

Comment:
is this not a critical failure here and you should be setting error  and returning false?

Reply:
I have a plan to support boot mode in the future, but I will make an error for boot mode at this stage to make it clear. Current target device do not support boot mode.